compiler/basicTypes/Var.hs

```diff
@@ -252,6 +254,7 @@ data Var
         realUnique :: {-# UNPACK #-} !Int,
         varType    :: Type,
         varMult    :: Mult,             -- See Note [Multiplicity of let binders]
+        varUsages  :: UsageAnnotation,
```

Comment:
This is a refactor which I have started, but it's more arduous that I would like, so I'd rather do it after merge. 

I don't know that we can do the special top level marker, because I don't think we fully control when things reach toplevel (and we may not have the opportunity to do a full substitution then).

Reply:
I'm fine without the top-level marker -- that was just because top-level let-bindings don't have anything interesting to store. But storing an empty `UsageAnnotation` still makes sense.

The problem with delaying the refactor is that it seems very difficult to get this all correct with the current structure. Maybe I'm wrong there, but for me this seems like more of an issue of correctness than just clean code.

Reply:
Indeed. In fact this is kind of the point: the forces us into much better compliance regarding classifying identifiers into let binders and lambda binders. Because the refactoring introduces a number of partial functions, it's easy to make GHC panic. Hence me being prudent here.

Reply:
Aha. So you're arguing for Python-like correctness ("it appears to work on the cases we've seen, even though there's not much science behind that, and doing *something* is better than just staring at type errors") instead of Haskell-like correctness ("it is correct"). :)

Of course, Haskell-like correctness is more expensive -- and that's your (valid) point.